Update getting started tutorial (#11221)
microsoft/FluidFramework: docs/content/docs/start/tutorial.md

2. [Creating a Fluid container](#creating-a-fluid-container).
3. [Writing the view](#writing-the-view).
4. [Connecting the view to Fluid data](#connecting-the-view-to-fluid-data).

All of the work in this demo will be done in the [app.js](https://github.com/microsoft/FluidHelloWorld/blob/main/src/app.js) file.

## Setting up the application

Start by creating a new instance of the Tinylicious client. Tinylicious is the Fluid Framework's local testing server, and a client is responsible for creating and loading containers.

Secondly, a constant is created to hold the key for a key-value pair that will represent the current value of the dice.

Next, the app creates Fluid containers using a schema that defines a set of *initial objects* that will be available in the container. In this case, the there is a single initial object of type [SharedMap][SharedMap]. You can think of a `SharedMap` as a set of key-value pairs similar to a JavaScript `Map` object. The "Shared" in the name conveys the fact that the object is accessible to code in every client simultaneously. Learn more about initial objects in [Data modeling]({{< relref "data-modeling.md" >}}).

> Lastly, `root` specifies the HTML element that renders the dice.

Author: I'll fix that.

    const { container } = await client.createContainer(containerSchema);
    // Set default data
    container.initialObjects.diceMap.set(diceValueKey, 1);
    // Attach container to service and return assigned ID
    const id = await container.attach();
    // Render the dice roller
    renderDiceRoller(container.initialObjects.diceMap, root);
    return id;
}
```

### Loading an existing container

Loading a container is more straightforward than creating a new one. When loading, the container already contains data, and is already attached, so those steps are irrelevant. You need only to pass the `id` of the container you wish to load in the `getContainer()` function along with the same schema used when creating the container. The schema is needed because the data in the container is compressed, so it needs to know the schema to reconstruct it.

```js
const loadExistingDice = async (id) => {
  const { container } = await client.getContainer(id, containerSchema);
  renderDiceRoller(container.initialObjects.diceMap, root);
}
```

### Switching between loading and creating

The application supports both creating a new container and loading an existing container using its `id`.

> Each client, at startup, needs to know if it is the first client, in which case it needs to create the container. Otherwise, simply loads an existing container. The sample app uses its own URL to find this out. The app stores the container ID in the URL hash.

Reviewer: ```suggestion
Each client, at startup, needs to know if it is the first client, in which case it needs to create the container.
Otherwise, it may simply load an existing container.
The sample app uses its own URL to find this out.
The app stores the container ID in the URL hash.
```

```js
const diceTemplate = document.createElement("template");

diceTemplate.innerHTML = `
  <style>
    .wrapper { text-align: center }
    .dice { font-size: 200px }
    .roll { font-size: 50px;}
  </style>
  <div class="wrapper">
    <div class="dice"></div>
    <button class="roll"> Roll </button>
  </div>
`
function renderDiceRoller(elem) {
    elem.appendChild(diceTemplate.content.cloneNode(true));
    const rollButton = elem.querySelector(".roll");
    const dice = elem.querySelector(".dice");

    rollButton.onclick = () => updateDice(Math.floor(Math.random() * 6)+1);

    const updateDice = (value) => {
        // Unicode 0x2680-0x2685 are the sides of a die (⚀⚁⚂⚃⚄⚅).
        dice.textContent = String.fromCodePoint(0x267f + value);
        dice.style.color = `hsl(${diceValue * 60}, 70%, 30%)`;
    };
    updateDice(1);
}
```

## Connecting the view to Fluid data

### Modifying Fluid data

> To begin using Fluid in the application, the first thing to change is what happens when the user clicks the `rollButton`. Instead of updating the local state directly, the button updates the number stored in the `value` key of the Fluid `diceMap` object that we created earlier. So, we added the `diceMap` as an additional parameter to the `rollButton` method.

Reviewer: The other step-by-step guides seem to be written in past tense, but this one is in present tense.

Author: Could you elaborate? I'm not sure what you mean. Would it help if the 2nd sentence said "Instead of updating the local state directly, the button **should update** the number stored in the `value` key of the Fluid `diceMap` object that we created earlier. "

Also, this isn't a step-by-step in the usual sense of having the reader take actions. (Step 1: do this. Step 2: do that.) It's really just a description of the app.js file, section-by-section. I'm not clear about how we would do that in the past tense.

Reviewer: Oh, sorry. I just mean in the other steps we say things like "Next, we needed to revise the `rollButton` method". I'm just wondering if this should be something like "To begin using Fluid in the application, the first thing we needed to change was...".

Author: I see. I'll fix that.

```js
const renderDiceRoller = (diceMap, elem) => {
```

Next, we needed to revise the `rollButton` method to assign the random dice value to the `diceMap` object.It does this by calling the `SharedMap.set` method. Because the `diceMap` is a Fluid `SharedMap`, changes will be distributed to all clients. Any changes to the `diceMap` will cause a `valueChanged` event to be emitted, and an event handler can trigger an update of the view.

This pattern is common in Fluid because it enables the view to behave the same way for both local and remote changes.

```js

>     rollButton.onclick = () => diceMap.set(diceValueKey, Math.floor(Math.random() * 6) + 1);

Reviewer: ```suggestion
rollButton.onclick = () => diceMap.set(diceValueKey, Math.floor(Math.random() * 6) + 1);
```

```js
    const updateDice = () => {
        const diceValue = diceMap.get(diceValueKey);
        // Unicode 0x2680-0x2685 are the sides of a dice (⚀⚁⚂⚃⚄⚅)
        dice.textContent = String.fromCodePoint(0x267f + diceValue);
        dice.style.color = `hsl(${diceValue * 60}, 70%, 30%)`;
    };
    updateDice();
```

### Handling remote changes

Now our local client updates the dice value indirectly by updating the `diceMap` object. Updating that object causes all changes made on this client to propagate to all other clients. But we need to get our local view to update whenever _any_ of the clients rolls the dice. To keep the view up to date as the data changes an event handler must be registered on the `diceMap` to call `updateDice` each time that the `valueChanged` event is sent from the Fluid relay service. So we added the following handler. See the [documentation for SharedMap][SharedMap] to get a list of events fired and the values passed to those events.

```js

>     diceMap.on("valueChanged", updateDice);

Reviewer: ```suggestion
diceMap.on("valueChanged", updateDice);
```

Reviewer: While we're here, if you don't mind :)
